### PR TITLE
add String#sub

### PR DESCRIPTION
**Pull request by jhass**
Like String#gsub, but with single replace semantics

There are a couple of usecases unsolved by String#gsub, but
maybe more common, if you know there's only a single match these
should be a tiny bit more performant

### CODE REVIEW

File: src/string.cr [hunk 1]
```diff
@@ -971,7 +971,156 @@ class String
...
-  # Returns a new string where each character yielded to the given block
+  # Returns a new string where the first character is yielded to the given
+  # block and replaced by its return value.
+  #
+  # ```
+  # "hello".sub {|x| (x.ord + 1).chr } #=> "iello"
+  # "hello".sub { "hi" } #=> "hiello"
+  # ```
+  def sub(&block : Char -> _)
+    return self if empty?
```
Comment (PragTob) [inline src/string.cr]: I believe a spec for this case is missing :)

Comment (PragTob) [inline src/string.cr]: the behaviour also seems not to be fully consistent through all variants of the method, e.g. ` def sub(hash : Hash(Char, _))` doesn't seem to return self when empty.

**Comment by PragTob** [comment]
Looks great to me, lots of test cases and good implementation - thanks :+1: 

Reading through just noticed one place where I was missing a spec :)

File: spec/std/string_spec.cr [hunk 2]
```diff
@@ -661,110 +661,215 @@ describe "String" do
...
-        "ex"
-      else
-        char
-      end
+      }
+      replaced.bytesize.should eq(9)
+      replaced.should eq("someoobar")
+
+      empty = ""
+      empty.sub { 'f' }.should be(empty)
```
Comment (jhass) [inline spec/std/string_spec.cr]: The spec for the empty case is here.

**Comment by jhass** [comment]
Returns `self` for the `empty?` string in `sub(hash : Hash(Char, _))` now, thanks for the catch.

File: spec/std/string_spec.cr (hunk 2 above)
Comment (PragTob) [inline spec/std/string_spec.cr]: ah sorry didn't see it as it was embedded in another spec - would prefer for it to be a separate test case :) But of course, that are details.

**Comment by PragTob** [comment]
:+1:

**Comment by asterite** [comment]
@jhass I thought so many times of implementing this but was very lazy to do it. I think a bunch of problems in Rosetta Code need this, and of course real programs would need this too.

Thank you so much!!! And thanks for the docs, specs, and existing doc fixes.

**Comment by jhass** [comment]
Same actually, it was itching me for some time but I never found the motivation until now :D

**Pull request merged**